### ISSUE TITLE
Renderer bug when mixing code blocks and math

### ISSUE THREAD

**Issue by kykc** [susam/texme#9]
````markdown
# TexMe Bug

```php
$profit = $sale->getProfit();
echo $profit;
```

$$
ax^2+bx+c=0
$$


```yml
metadata:
  tags:
    - texme
    - bug
```
````
Such markdown snippet will be rendered to 
````html
<h1>TexMe Bug</h1>
<pre>
    <code class="language-php">
        $profit = $sale-&gt;getProfit();
        echo $profit;
```
$$
ax^2+bx+c=0
$$

```yml
metadata:
  tags:
    - texme
    - bug
</code></pre>
````
While it should be rendered like this
```html
<h1>TexMe Bug</h1>
<pre>
    <code class="language-php">
        $profit = $sale-&gt;getProfit();
        echo $profit;
    </code>
</pre>

$$
ax^2+bx+c=0
$$


<pre>
    <code class="language-yml">
metadata:
  tags:
    - texme
    - bug
    </code>
</pre>
```
Seems like a greedy regex issue or something like that.
Rendering was done from `node.js` side like this:
```js
var texme = require('texme')
var input = ''; // snipped code
var output = texme.render(input)
console.log(output)
```

**Comment by bachmeil** [comment]
Same problem here. It chokes on any code that contains dollar signs. I inserted `&#36;` inside my code blocks, and for some reason that is converted to a dollar sign, then causes it to choke. Something is going on that causes it to not respect code blocks (no transformation should be taking place inside a code block).

**Comment by bachmeil** [comment]
Adding

```
<script>window.texme = { protectMath: false }</script>
```

does fix the problem, though I don't know what that might break. It's unclear what `protectMath` does; I haven't had any problems rendering documents with that option set to false.

**Comment by bachmeil** [comment]
There is no easy fix that will make this work as expected for all valid documents. It would be necessary for texme to ignore all code blocks, and that would require a full markdown parser. Commonmark is not suited for documents that contain equations because it has no way to selectively turn off processing of text. Any notation used to add an equation in the text could potentially appear in a code block.

A hack is to use notation to add equations to the text that will never be used in a code block. Once you do that, you are no longer supporting Commonmark documents.

**Comment by susam** [comment]
Thank you for reporting and investigating this issue, @kykc and @bachmeil. I have added support for a special purpose environment named `md` that protects LaTeX delimiters (`$`, `$$`, etc.) within Markdown code blocks and code spans. With this new environment, the Markdown code in the original bug report by @kykc can be rewritten as follows to fix the rendering issue:

````markdown
<!DOCTYPE html><script src="https://cdn.jsdelivr.net/npm/texme@0.9.0"></script><textarea>

# Demo

\begin{md}
```php
$profit = $sale->getProfit();
echo $profit;
```
\end{md}

$$
ax^2+bx+c=0
$$


```yml
metadata:
  tags:
    - texme
    - bug
```
````

Here is how the output looks: <https://jsfiddle.net/zwarLyv5/>. This feature is available since version 0.9.0. See [README: Markdown Priority Environment](https://github.com/susam/texme/blob/6213f86/README.md#markdown-priority-environment) for more details on this.
